specs/bellatrix/fork-choice.md

```diff
@@ -76,6 +77,102 @@ As per EIP-3675, before a post-transition block is finalized, `notify_forkchoice
+being re-orged. Complete information for evaluating `get_proposer_head` _will not_ be available
+immediately after the receipt of a new block, so an approximation of those conditions should be
+used when deciding whether to send or suppress a fork choice notification. The exact conditions
+used may be implementation-specific, a suggested implementation is below.
+
+Let `validator_is_connected` be a function that indicates whether the validator with
+`validator_index` is connected to the node (e.g. has sent an unexpired proposer preparation
+message).
+
+```python
+def validator_is_connected(validator_index: ValidatorIndex) -> boolean:
+    ...
+```
+
+```python
+def should_override_forkchoice_update(
+    store: Store,
+    head_root: Root,
+) -> boolean:
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    head_block = store.blocks[head_root]
+    parent_root = head_block.parent_root
+    parent_block = store.blocks[parent_root]
+    current_slot = get_current_slot(store)
+    proposal_slot = head_block.slot + 1
+
+    # Only re-org the head_block block if it arrived later than the attestation deadline.
+    head_late = store.block_timeliness.get(head_root) is False
+
+    # Only suppress the fork choice update if we are confident that we will propose the next block.
+    parent_state_advanced = store.block_states[parent_root]
+    process_slots(parent_state_advanced, proposal_slot)
+    proposer_index = get_beacon_proposer_index(parent_state_advanced)
+    proposing_reorg_slot = validator_is_connected(proposer_index)
+
+    # Do not re-org if the chain is not finalizing with acceptable frequency.
+    proposal_epoch = compute_epoch_at_slot(proposal_slot)
+    epochs_since_finalization = proposal_epoch - store.finalized_checkpoint.epoch
+    finalization_ok = epochs_since_finalization <= REORG_MAX_EPOCHS_SINCE_FINALIZATION
+
+    # Single slot re-org.
+    parent_slot_ok = parent_block.slot + 1 == head_block.slot
+    time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT
+    current_time_ok = (head_block.slot == current_slot or
+                       (proposal_slot == current_slot and
+                        time_into_slot <= SECONDS_PER_SLOT // INTERVALS_PER_SLOT // 2))
+    single_slot_reorg = parent_slot_ok and current_time_ok
+
+    # Shuffling stable.
+    shuffling_stable = proposal_slot % SLOTS_PER_EPOCH != 0
+
+    # FFG information of the new head_block will be competitive with the current head.
+    # FIXME: needs more justification/finalization information in the store
+    # ffg_competitive =
+    #    store.realized_justifications[parent_root] ==
+    #        store.realized_justifications[head_root] and
+    #    store.realized_finalizations[parent_root] ==
+    #        store.realized_finalization[head_root]
+    ffg_competitive = True
+
+    # Check the head weight only if the attestations from the head slot have already been applied.
```

Comment:
```suggestion
    # Check the head weight only if the attestations from the current slot have already been applied.
```
This function is phrased in a way that it is agnostic to whether we are calling it on the slot we are about to propose or during the previous slot. However, what worries me is the situation where `current_slot = N`, we are supposed to propose at `N+1`, `head_slot = N-1` and we call during slot `N`. In this case, suppose the block `N` is  based on `N-2`, forking off `N-1`, it is a timely block, and we haven't yet counted attestations because we are calling this function during `N`. What will happen is that our head will still be `N-1` and every check here will pass. It may well be that later, at the time we need to propose, during N+1, this timely block at N  would become our head, but this function here may return true.  

Now this is not a problem per-se in this specification since later on, when calling `get_proposer_head` we will get a misprediction because we will not have passed `single_slot_reorg`.  Anyway, I feel it's better to change this condition here from checking that the attestations for the head block to have been counted to "changing that the attestations for the current slot have been counted if we are calling late in the slot". I do recognize that this is difficult to specify, but perhaps it's better to not be agnostic as to the timing in which this function is being called. 

As a side in this long comment: if we are in the situation above, and we call this function during N, then the `assert store.proposer_boost_root == Root()` will fail in this function.

Reply:
If the head is at `N - 1` and we are proposing at `N + 1` then this function won't return true. This function infers the `proposal_slot` _from_ the head slot, so it will check whether we are proposing at slot _N_:

https://github.com/ethereum/consensus-specs/blob/56f20f7214ca17ef87cbea416181f9b05f944b1d/specs/bellatrix/fork-choice.md?plain=1#L111

i.e. if we are actually proposing at `N+1` and some other block for `N` already exists, the `proposing_reorg_slot` check will be `False`.

Therefore I think the spec already minimises the chance of a misprediction. It's a really nice test case though, I'll make sure it gets a spec test when we add them (I'll also add a test to Lighthouse before then).

> Anyway, I feel it's better to change this condition here from checking that the attestations for the head block to have been counted to "changing that the attestations for the current slot have been counted if we are calling late in the slot".

I'm open to making this change, but would like to clarify something about how you envisage Prysm's implementation applying attestations at 10 or 11 secs into the slot will work: are you going to avoid updating the fork choice store's current slot "early"? I.e. if you are 10s into slot `N` will you apply the attestations for slot `N` and leave `get_current_slot(store) == N`? The way everything is written at the moment maintains the invariant that only attestations from `get_current_slot(store) - 1` or earlier are applied, supporting Lighthouse's implementation which advances the fork choice slot early at 11.5s into slot `N` and pre-emptively increments the fork choice slot to `N + 1`.

> As a side in this long comment: if we are in the situation above, and we call this function during N, then the assert store.proposer_boost_root == Root() will fail in this function.

Nice catch on the `assert`! I think we need to change it so we return `False` rather than erroring, or assert that the proposer boost is for a block other than `head_block`:

```python
assert store.proposer_root_root != head_root
```

Reply:
> i.e. if we are actually proposing at N+1 and some other block for N already exists, the `proposing_reorg_slot` check will be `False`.
Indeed I mixed the previous check, but now I'm worried about something else: `validator_is_connected` can return true for a large number of validators in the case of a pool. In the above scenario, a pool proposing at N and N+1 will actually pass with the wrong validator the check for `proposing_reorg_slot`. 

>  are you going to avoid updating the fork choice store's current slot "early"? I.e. if you are 10s into slot N will you apply the attestations for slot N and leave `get_current_slot(store) == N` ?
Indeed that's my plan, I'm open to change this to make it closer to LightHouse, but it seems very hard on our end cause we track the wallclock on forkchoice and tricking it to change the slot involves bad hacks like changing the genesis time.

Reply:
> validator_is_connected can return true for a large number of validators in the case of a pool. In the above scenario, a pool proposing at N and N+1 will actually pass with the wrong validator the check for proposing_reorg_slot.

@potuz I don't understand what you mean here, can you spell it out in a bit more detail? I thought we already covered this case by demonstrating that the fcU for `N - 1` won't be suppressed, and if the head switches to `N` upon counting attestations towards the end of slot `N`, then the fcU for `N` _also_ won't be suppressed because it will fail the `parent_slot_ok` check (`N`'s parent is `N - 2`). Hence we'll still build on `N` via the normal process (no re-org attempts).